### PR TITLE
82 login backend

### PR DESCRIPTION
**Pull request by husamkamal**

### CODE REVIEW

**Review by mai-mohammed — APPROVED** [review]

File: server/src/helpers/passwordCompare.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import bcrypt from 'bcryptjs';
+
+const passwordCompare = (password:string, hash:string) => {
+  const isTrue = new Promise((resolve, reject) => {
+    bcrypt.compare(password, hash, (err, res) => {
```
Comment (LinaYahya) [inline server/src/helpers/passwordCompare.ts]:
```suggestion
    bcrypt.compare(password, hash, (err, res) => {
```
// no need to convert to promise as bcrypt supports promise by nature

File: server/src/helpers/generateToken.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import jwt from 'jsonwebtoken';
+
+const generateToken = (payload) => {
+  const tokenGenerate = new Promise((resolve, reject) => {
+    jwt.sign(
+      { role: 'user', payload },
```
Comment (LinaYahya) [inline server/src/helpers/generateToken.ts]:
```suggestion
      payload,
```
SO we can use `generateToken` to both admin and user auth

File: server/src/helpers/generateToken.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import jwt from 'jsonwebtoken';
+
+const generateToken = (payload) => {
+  const tokenGenerate = new Promise((resolve, reject) => {
```
Comment (LinaYahya) [inline server/src/helpers/generateToken.ts]:
```suggestion
const generateToken = (payload) => new Promise((resolve, reject) => {
```
return in one line

File: server/src/services/login.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import { Customer } from '../db/models';
+
+const loginServices = async ({ email }) => {
```
Comment (LinaYahya) [inline server/src/services/login.ts]: no need to create a service, or use descriptive meaning

File: server/src/controllers/login.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import { Request } from 'express';
+import { loginSchema } from '../validation';
+import { loginServices } from '../services';
+import { generateToken, passwordCompare } from '../helpers';
+
+const loginController = async (req:Request) => {
+  const { email, password } = req.body;
+  await loginSchema.validate({ email, password });
+  const result:{ password:string, id:number, fullName:string } = await loginServices({ email });
+  if (!result?.password.length) {
```
Comment (LinaYahya) [inline server/src/controllers/login.ts]:
```suggestion
  if (!result) { // why to check on password specifically, result will be null if there's no user 
```

File: server/src/controllers/login.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    return { status: 200, msg: 'done', data: { msg: 'user not found' } };
```
Comment (LinaYahya) [inline server/src/controllers/login.ts]:
```suggestion
    throw createErr(400, )
```

**Review by LinaYahya — CHANGES_REQUESTED** [review]

**Review by LinaYahya — APPROVED** [review]

**Review by LinaYahya — APPROVED** [review]

**Review by abdou059 — APPROVED** [review]

**Pull request merged**